File: shareholders_management/shareholders/report/account_statement/account_statement.js
// Copyright (c) 2016, Ahmad Al-Farran and contributors
// For license information, please see license.txt
/* eslint-disable */
var from_date = undefined;
var to_date = undefined;
frappe.query_reports["Account statement"] = {
	"filters": [
		{
			"fieldname": "shareholder_name",
			"label": __("Shareholder Name"),
			"fieldtype": "Link",
			"options": "Shareholder",
			"on_change": function(query_report) {
				var shareholder_name = query_report.get_values().shareholder_name;
				frappe.query_report.set_filter_value("shareholder_account","");
			},
			"reqd": 1
		},
		{
			"fieldname": "shareholder_account",
			"label": __("Shareholder Account"),
			"fieldtype": "Link",
			"options": "Shareholder Account",
			"get_query": function() {
				var shareholder_name = frappe.query_report.get_filter_value('shareholder_name');
				return {
					"doctype": "Shareholder Account",
					"filters": {
						"shareholder_name": shareholder_name,
					}
				}
			},
			"on_change": function(query_report) {
				var shareholder_account = query_report.get_values().shareholder_account;
				if (!shareholder_account) {
					frappe.query_report.set_filter_value("from_date", "");
					frappe.query_report.set_filter_value("to_date", "");
					frappe.query_report.set_filter_value("currency", "");
					return;
				}
				frappe.call({
					method: "shareholders_management.shareholders.report.account_statement.account_statement.all_operations_dates",
					args: {
						parent: query_report.get_filter_value("shareholder_account"),
						operation: query_report.get_filter_value("operation")
					},
					callback: function(r) {
						from_date = r.message[0].from_date;
						frappe.query_report.set_filter_value("from_date", from_date);
						to_date = r.message[0].to_date;
						frappe.query_report.set_filter_value("to_date", to_date);
					}
				});
				frappe.model.with_doc("Shareholder Account", shareholder_account, function(r) {
					var sa = frappe.model.get_doc("Shareholder Account", shareholder_account);
					frappe.query_report.set_filter_value("currency", sa.currency);
				});
			},
			"reqd": 1
		},
		{
			"fieldname": "operation",
			"label": __("Operation Type"),
			"fieldtype": "Select",
			"options": __("All Operations") + "\n" + __("Deposit") + "\n" + __("Withdrawal"),
			"default": __("All Operations"),
			"on_change": function(query_report) {
				frappe.call({
					method: "shareholders_management.shareholders.report.account_statement.account_statement.all_operations_dates",
					args: {
						parent: query_report.get_filter_value("shareholder_account"),
						operation: query_report.get_filter_value("operation")
					},
					callback: function(r) {
						from_date = r.message[0].from_date;
						frappe.query_report.set_filter_value("from_date", from_date);
						to_date = r.message[0].to_date;
						frappe.query_report.set_filter_value("to_date", to_date);
					}
				});
			}
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"on_change": function(query_report) {
				var f_date = query_report.get_filter_value("from_date");
				var t_date = query_report.get_filter_value("to_date");
				if (!f_date) {
					frappe.query_report.refresh();
					return;
				}
				if (f_date > t_date) {
					frappe.query_report.set_filter_value("from_date", from_date);
					frappe.throw(__("You can not enter a date after {0}!!",[t_date]))
				}
				frappe.query_report.refresh();
			},
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"on_change": function(query_report) {
				var f_date = query_report.get_values().from_date;
				var t_date = query_report.get_values().to_date;
				if (!t_date) {
					frappe.query_report.refresh();
					return;
				}
				if (t_date < f_date) {
					frappe.query_report.set_filter_value("to_date", to_date);
					frappe.throw(__("You can not enter a date before {0}!!",[f_date]))
				}
				frappe.query_report.refresh();
			},
		},
		{
			"fieldname": "currency",
			"label": __("Currency"),
			"fieldtype": "Link",
			"options": "Currency",
			"read_only": 1
		},
	]
};

File: shareholders_management/shareholders/report/account_statement/account_statement.py
# Copyright (c) 2013, Ahmad Al-Farran and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
# from erpnext.projects.doctype.log_frame.log_frame import get_logframe, get_child_logframe, get_add_output_to_the_main_goal
from six import BytesIO
from frappe.utils import getdate
# from docxtpl import DocxTemplate
import os


def execute(filters=None):
    # validate_filters(filters)
    filters = frappe._dict(filters or {})
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data


# def validate_filters(filters):
#     filters.from_date = getdate(filters.from_date)
#     filters.to_date = getdate(filters.to_date)
#     if filters.from_date and filters.to_date:
#         if filters.from_date > filters.to_date:
#             frappe.throw(_("'From Date' cannot be greater than 'To Date'"))


def get_columns(filters):
    columns = [
        {
            "label": _("Date"),
            "fieldtype": "Date",
            "fieldname": "date",
            "width": '110'
        },
        {
            "label": _("Operation Type"),
            "fieldtype": "Data",
            "fieldname": "operation",
            "width": '100'
        },
        {
            "label": _("Amount"),
            "fieldtype": "Currency",
            "fieldname": "amount",
            "width": '170'
        },
        {
            "label": _("Operation Purpose"),
            "fieldtype": "Data",
            "fieldname": "purpose",
            "width": '170'
        },
        {
            "label": _("Note"),
            "fieldtype": "Data",
            "fieldname": "description",
            "width": '800',
        },
    ]
    return columns


def get_conditions(filters):
    conditions = {}
    if filters.shareholder_account and filters.from_date and filters.to_date:
        conditions["operation"] = filters.operation
        conditions["shareholder_account"] = filters.shareholder_account
        conditions["from_date"] = filters.from_date
        conditions["to_date"] = filters.to_date
    return conditions


def get_data(filters):
    data = []
    conditions = get_conditions(filters)
    if conditions:
        if conditions["operation"] == _("Deposit"):
            operations = frappe.db.sql("""
        		SELECT deposit_date as op_date, parentfield as op, amount, deposit_type as op_type, description
        		FROM `tabDeposit`
        		WHERE parent = %s
        		AND deposit_date BETWEEN %s AND %s
        		ORDER BY deposit_date""", (conditions["shareholder_account"], conditions["from_date"], conditions["to_date"]), as_dict=True)
        elif conditions["operation"] == _("Withdrawal"):
            operations = frappe.db.sql("""
        		SELECT withdrawal_date as op_date, parentfield as opr, amount, withdrawal_type as op_type, description
        		FROM `tabWithdrawal`
        		WHERE parent = %s
        		AND withdrawal_date BETWEEN %s AND %s
        		ORDER BY withdrawal_date""", (conditions["shareholder_account"], conditions["from_date"], conditions["to_date"]), as_dict=True)
        else:
            operations = frappe.db.sql("""
        		SELECT withdrawal_date as op_date, parentfield as opr, amount, withdrawal_type as op_type, description
        		FROM `tabWithdrawal`
        		WHERE parent = %(parent)s
        		AND withdrawal_date BETWEEN %(from)s AND %(to)s
        		UNION
        		SELECT deposit_date as op_date, parentfield as op, amount, deposit_type as op_type, description
        		FROM `tabDeposit`
        		WHERE parent = %(parent)s
        		AND deposit_date BETWEEN %(from)s AND %(to)s
        		ORDER BY op_date""", {"parent": conditions["shareholder_account"],
                                "from": conditions["from_date"],
                                "to": conditions["to_date"]}, as_dict=True)

        if operations:
            for i in operations:
                opr = _("Deposit")
                if i.opr == 'withdrawal_operations':
                    opr = _("Withdrawal")
                data.append({"date": i.op_date, "operation": opr, "amount": i.amount,
                             "purpose": _(i.op_type), "description": i.description})
    return data

@frappe.whitelist()
def all_operations_dates(parent, operation):
    dates = []
    if operation == _("Deposit"):
        data = frappe.db.sql("""
        SELECT min(deposit_date) as from_date, max(deposit_date) as to_date
        FROM `tabDeposit`
        WHERE parent = %(parent)s""", {"parent": parent}, as_dict=True)
    elif operation == _("Withdrawal"):
        data = frappe.db.sql("""
        SELECT min(withdrawal_date) as from_date, max(withdrawal_date) as to_date
        FROM `tabWithdrawal`
        WHERE parent = %(parent)s""", {"parent": parent}, as_dict=True)
    else:
        data = frappe.db.sql("""
        SELECT min(o_dates.from_date) as from_date, max(o_dates.to_date) as to_date
        FROM (
        SELECT min(deposit_date) as from_date, max(deposit_date) as to_date
        FROM `tabDeposit`
        WHERE parent = %(parent)s
        UNION
        SELECT min(withdrawal_date) as from_date, max(withdrawal_date) as to_date
        FROM `tabWithdrawal`
        WHERE parent = %(parent)s) as o_dates""", {"parent": parent}, as_dict=True)
    for i  in data:
        dates.append({"from_date": i.from_date, "to_date": i.to_date})
    return dates
